JavaScript/src/dateTime/baseHoliday.ts

```diff
@@ -0,0 +1,247 @@
+            hasYear = true;
+        }
+        else {
+            year = referenceDate.getFullYear();
+        }
+
+        let holidayKey: string;
+        for (holidayKey of this.config.holidayNames.keys()) {
+            if (this.config.holidayNames.get(holidayKey).indexOf(holidayStr) > -1) {
+                break;
+            }
+        }
+
+        if (holidayKey) {
+            let timexStr: string;
+            let value = referenceDate;
+            let func = this.config.holidayFuncDictionary.get(holidayKey);
+            if (func) {
+                value = func(year);
+                timexStr = this.config.variableHolidaysTimexDictionary.get(holidayKey);
+                if (!timexStr) {
+                    timexStr = `-${FormatUtil.toString(value.getMonth() + 1, 2)}-${FormatUtil.toString(value.getDate(), 2)}`;
+                }
+            }
+            else {
+                return ret;
+            }
+
+            if (hasYear) {
+                ret.timex = FormatUtil.toString(year, 4) + timexStr;
+                ret.futureValue = ret.pastValue = new Date(year, value.getMonth(), value.getDate());
+                ret.success = true;
+                return ret;
+            }
+
+            ret.timex = "XXXX" + timexStr;
+            ret.futureValue = this.getFutureValue(value, referenceDate, holidayKey);
+            ret.pastValue = this.getPastValue(value, referenceDate, holidayKey);
+            ret.success = true;
+
+            return ret;
+        }
+
+        return ret;
+    }
+
+    private getFutureValue(value: Date, referenceDate: Date, holiday: string): Date {
+        if (value < referenceDate) {
+            let func = this.config.holidayFuncDictionary.get(holiday);
+            if (func) {
+                return func(value.getFullYear() + 1);
+            }
+        }
+        return value;
+    }
+
+    private getPastValue(value: Date, referenceDate: Date, holiday: string): Date {
+        if (value >= referenceDate) {
+            let func = this.config.holidayFuncDictionary.get(holiday);
+            if (func) {
+                return func(value.getFullYear() - 1);
+            }
+        }
+        return value;
+    }
+}
+
+export abstract class BaseHolidayParserConfiguration implements IHolidayParserConfiguration {
+    variableHolidaysTimexDictionary: ReadonlyMap<string, string>;
+    holidayFuncDictionary: ReadonlyMap<string, (year: number) => Date>;
+    holidayNames: ReadonlyMap<string, string[]>;
+    holidayRegexList: RegExp[];
+    abstract getSwiftYear(text: string): number;
+    abstract sanitizeHolidayToken(holiday: string): string;
+
+    constructor() {
+        this.variableHolidaysTimexDictionary = BaseDateTime.VariableHolidaysTimexDictionary;
+        this.holidayFuncDictionary = this.initHolidayFuncs();
+    }
+
+    protected initHolidayFuncs(): ReadonlyMap<string, (year: number) => Date> {
+        return new Map<string, (year: number) => Date>(
+            [
```

Comment:
We should think of a way to make this auto-generated in the future.
Can you add a TODO here?